File: java/src/main/java/falcon/api/command/CommandContext.java
package falcon.api.command;

import falcon.api.Player;
import falcon.api.internal.Interop;
import falcon.api.internal.PluginContext;

import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.util.List;
import java.util.Optional;

public final class CommandContext {
    private final PluginContext mContext;
    private final MemorySegment mSender;
    private final List<String> mArguments;

    CommandContext(PluginContext context, MemorySegment sender, List<String> arguments) {
        mContext = context;
        mSender = sender;
        mArguments = List.copyOf(arguments);
    }

    public String senderName() {
        return Interop.string(Interop.api().senderName(mSender));
    }

    public Optional<Player> player() {
        MemorySegment player = Interop.api().senderPlayer(mSender);
        if (Interop.isNull(player)) {
            return Optional.empty();
        }
        return Optional.of(new Player(mContext, player));
    }

    public List<String> arguments() {
        return mArguments;
    }

    public boolean hasPermission(String node) {
        try (Arena arena = Arena.ofConfined()) {
            return Interop.api().senderHasPermission(mSender, Interop.text(arena, node)) != 0;
        }
    }

    public void reply(String message) {
        try (Arena arena = Arena.ofConfined()) {
            Interop.api().senderSendMessage(mSender, Interop.text(arena, message));
        }
    }
}

File: dotnet/Falcon.PluginAPI/Commands/CommandContext.cs
using System.Collections.Generic;
using Falcon.Interop;

namespace Falcon
{
    public sealed unsafe class CommandContext
    {
        private readonly FalconCommandSender* _sender;

        internal CommandContext(FalconCommandSender* sender, IReadOnlyList<string> arguments)
        {
            _sender = sender;
            Arguments = arguments;
        }

        public IReadOnlyList<string> Arguments { get; }

        public nint Handle
        {
            get
            {
                return (nint)_sender;
            }
        }

        public string SenderName
        {
            get
            {
                return NativeApi.Text(NativeApi.Table->senderName(_sender));
            }
        }

        public Player? Player
        {
            get
            {
                return Player.Wrap(NativeApi.Table->senderPlayer(_sender));
            }
        }

        public bool HasPermission(string node)
        {
            fixed (byte* text = NativeApi.Utf8(node))
            {
                return NativeApi.Table->senderHasPermission(_sender, text) != 0;
            }
        }

        public void Reply(string message)
        {
            fixed (byte* text = NativeApi.Utf8(message))
            {
                NativeApi.Table->senderSendMessage(_sender, text);
            }
        }
    }
}
